### PR TITLE
Fix pyroot on Windows

### PR DESCRIPTION
**Pull request by bellenot**
- RootMacros.cmake: set the correct PYTHONPATH (${ROOTSYS}/bin on Windows)
 - CPyCppyy/CMakeLists.txt: export all symbols and set correct prefix (lib) and suffix (.pyd)
 - MemoryRegulator[.cxx/.h]: downstream cppyy changes [`move hook setting to .cxx`](https://bitbucket.org/wlav/cpycppyy/commits/24ac364)
 - cppyy-backend/CMakeLists.txt: export all symbols
 - pythonizations/CMakeLists.txt: export all symbols, set correct suffix (.pyd) and link against Python library (cannot have unresolved symbols on Windows)
 - PyzPythonHelpers.cxx, RDataFramePyz.cxx, RVecPyz.cxx, TPyClassGenerator.cxx, and TPython.cxx: Use proper forward declarations and fix pointer formatting
 - pythonizations/test/CMakeLists.txt: disable failing tests

### CODE REVIEW

File: bindings/pyroot/cppyy/CPyCppyy/CMakeLists.txt [hunk 1]
```diff
@@ -60,13 +60,18 @@ foreach(val RANGE ${how_many_pythons})
     target_link_libraries(${libname} PUBLIC -Wl,-bind_at_load -Wl,-undefined -Wl,dynamic_lookup cppyy_backend${python_under_version_string})
   elseif(NOT MSVC)
     target_link_libraries(${libname} PUBLIC -Wl,--unresolved-symbols=ignore-all cppyy_backend${python_under_version_string})
+  else()
```
Comment (etejedor) [inline bindings/pyroot/cppyy/CPyCppyy/CMakeLists.txt:63]: Just a matter of taste, but I would do:
`if(APPLE)
...
elseif(MSVC)
...
else
...
```
Looks cleaner to me.

File: bindings/pyroot/cppyy/CPyCppyy/CMakeLists.txt [hunk 2]
```diff
@@ -93,6 +98,7 @@ foreach(val RANGE ${how_many_pythons})
   target_include_directories(${libname}
      PRIVATE
         ${CMAKE_SOURCE_DIR}/core/foundation/inc   # needed for string_view backport
+        ${CMAKE_BINARY_DIR}/ginclude
```
Comment (etejedor) [inline bindings/pyroot/cppyy/CPyCppyy/CMakeLists.txt:101]: Is this only needed for Windows? It does not interfere in the other platforms?

File: bindings/pyroot/cppyy/CPyCppyy/src/MemoryRegulator.cxx [hunk 3]
```diff
@@ -244,3 +244,16 @@ PyObject* CPyCppyy::MemoryRegulator::RetrievePyObject(Cppyy::TCppObject_t cppobj
 
     return nullptr;
 }
+
+
+//-----------------------------------------------------------------------------
+void CPyCppyy::MemoryRegulator::SetRegisterHook(MemHook_t h) {
```
Comment (etejedor) [inline bindings/pyroot/cppyy/CPyCppyy/src/MemoryRegulator.cxx:250]: I saw these changes (and the ones in the header) are now in upstream cppyy, nice!

I believe this is the only thing you needed to contribute with upstream?

File: bindings/pyroot/cppyy/CPyCppyy/CMakeLists.txt (hunk 2 above)
Comment (bellenot) [reply bindings/pyroot/cppyy/CPyCppyy/CMakeLists.txt:101]: > Is this only needed for Windows? It does not interfere in the other platforms?

I don't think so, the PR build (once it works) should show us if it has side effects

File: bindings/pyroot/cppyy/CPyCppyy/src/MemoryRegulator.cxx (hunk 3 above)
Comment (bellenot) [reply bindings/pyroot/cppyy/CPyCppyy/src/MemoryRegulator.cxx:250]: > I saw these changes (and the ones in the header) are now in upstream cppyy, nice!
> 
> I believe this is the only thing you needed to contribute with upstream?

Thanks! And yes, it's the only requirement I had

File: bindings/pyroot/cppyy/CPyCppyy/CMakeLists.txt (hunk 1 above)
Comment (bellenot) [reply bindings/pyroot/cppyy/CPyCppyy/CMakeLists.txt:63]: As you wish (I simply added the `else()`)

File: bindings/pyroot/pythonizations/CMakeLists.txt [hunk 4]
```diff
@@ -95,6 +95,11 @@ foreach(val RANGE ${how_many_pythons})
   add_library(${libname} SHARED ${cpp_sources})
   # Set the suffix to '.so' and the prefix to 'lib'
   set_target_properties(${libname} PROPERTIES  ${ROOT_LIBRARY_PROPERTIES})
+  if(MSVC)
```
Comment (etejedor) [inline bindings/pyroot/pythonizations/CMakeLists.txt:98]: And the same rational could be applied here:
```
if(APPLE)
elseif(MSVC)
else
```

File: bindings/pyroot/pythonizations/src/PyzPythonHelpers.cxx [hunk 5]
```diff
@@ -29,6 +28,12 @@ PyROOT extension module.
 
 #include <sstream>
 
+namespace CPyCppyy {
```
Comment (etejedor) [inline bindings/pyroot/pythonizations/src/PyzPythonHelpers.cxx:31]: Can you add a comment that this is needed for Windows?
It should be fine in the other platforms as well?

File: bindings/tpython/src/TPyClassGenerator.cxx [hunk 6]
```diff
@@ -29,6 +27,13 @@
 #include <string>
 #include <typeinfo>
 
+namespace CPyCppyy {
```
Comment (etejedor) [inline bindings/tpython/src/TPyClassGenerator.cxx:30]: Also here

File: bindings/pyroot/pythonizations/src/PyzPythonHelpers.cxx (hunk 5 above)
Comment (bellenot) [reply bindings/pyroot/pythonizations/src/PyzPythonHelpers.cxx:31]: Yes, it's simply a forward declaration

File: bindings/tpython/src/TPython.cxx [hunk 7]
```diff
@@ -97,7 +96,13 @@ ClassImp(TPython);
 static PyObject *gMainDict = 0;
 
 namespace CPyCppyy {
-    extern PyObject *gThisModule;
+   R__EXTERN PyObject *gThisModule;
+   namespace PyStrings {
```
Comment (etejedor) [inline bindings/tpython/src/TPython.cxx:100]: And  here

File: bindings/pyroot/pythonizations/CMakeLists.txt (hunk 4 above)
Comment (bellenot) [reply bindings/pyroot/pythonizations/CMakeLists.txt:98]: No problem

File: cmake/modules/RootMacros.cmake [hunk 8]
```diff
@@ -1728,10 +1733,15 @@ endfunction()
 #----------------------------------------------------------------------------
 function(ROOT_ADD_PYUNITTEST name file)
   CMAKE_PARSE_ARGUMENTS(ARG "WILLFAIL" "" "COPY_TO_BUILDDIR;ENVIRONMENT;DEPENDENCIES_FOUND" ${ARGN})
-  set(ROOT_ENV ROOTSYS=${ROOTSYS}
-      PATH=${ROOTSYS}/bin:$ENV{PATH}
-      LD_LIBRARY_PATH=${ROOTSYS}/lib:$ENV{LD_LIBRARY_PATH}
-      PYTHONPATH=${ROOTSYS}/lib:$ENV{PYTHONPATH})
+  if(MSVC)
```
Comment (etejedor) [inline cmake/modules/RootMacros.cmake:1736]: Nice, thanks for adapting this after we removed the remnant of the python subdirectories.

**Pull request merged**